lib/rust/parser/src/source/code.rs

```diff
@@ -205,17 +240,37 @@ impl<'s> AddAssign<&Code<'s>> for Code<'s> {
 /// The length of a [`Code`] object.
 #[derive(Copy, Clone, Debug, Default, Eq, PartialEq, Serialize, Reflect, Deserialize)]
 pub struct Length {
+    /// An offset, in UTF-8 code units (bytes).
     #[reflect(skip)]
     #[serde(skip)]
-    utf8:  usize,
-    utf16: u32,
+    pub utf8:         u32,
+    /// An offset, in UTF-16 code units (two-byte words).
+    pub utf16:        u32,
+    /// A difference in line numbers.
+    #[reflect(hide)]
+    pub newlines:     u32,
+    /// If `newlines` is 0, this is the difference in UTF-16 code-unit positions within a line; if
+    /// `newlines` is nonzero, this is the position within the line ending the range.
+    pub line_chars16: u32,
 }
 
 impl Length {
     /// Returns the length of the given input.
     #[inline(always)]
     pub fn of(s: &str) -> Self {
-        Self { utf8: s.len(), utf16: s.encode_utf16().count() as u32 }
+        let mut utf16 = 0;
+        let mut newlines = 0;
+        let mut line_chars16 = 0;
+        for c in s.chars() {
+            let char_len16 = c.len_utf16() as u32;
+            utf16 += char_len16;
+            line_chars16 += char_len16;
+            if c == '\n' {
+                newlines += 1;
+                line_chars16 = 0;
+            }
```

Comment:
Somewhere else we say about interpreting CRLF, CR and LF as newlines, but here we consider only LF.

Reply:
Fixed and added testing.